File: ai-mock/ai/state.py
from queue import PriorityQueue

from ai import *


class State:
    def __init__(self):
        self.field = []
        self.ninjas = [None, None]
        self.dogs = {}
        self.dog_points = set()
        self.power = 0
        self.souls = set()
        self.skills = []
        self.exceptions = set()
        self.steps = [[], []]
        self.doppelganger = None
        self._prev_ninjas = None
        self._steps_from_ninja = None
        self._steps_from_doppelganger = None
        self._prev_doppel_point = None

    @property
    def steps_from_ninja(self):
        """
        忍者から各マスへの最短経路
        """
        ninjas_hash = hash(self.ninjas[0]) + hash(self.ninjas[1]) * 2
        if (not self._prev_ninjas or 
                ninjas_hash != self._prev_ninjas):
            self._prev_ninjas = ninjas_hash
            self._steps_from_ninja = self.steps_from_points([
                self.ninjas[0].point,
                self.ninjas[1].point
            ])
        return self._steps_from_ninja
    
    @property
    def steps_from_doppelganger(self):
        """
        影武者から各マスへの距離
        """
        if not self.doppelganger:
            return self.steps_from_ninja
        if not self._prev_doppel_point or self._prev_doppel_point != self.doppelganger:
            self._prev_doppel_point = self.doppelganger
            self._steps_from_doppelganger = self.steps_from_points([self.doppelganger])
        return self._steps_from_doppelganger
            

    def start(self):
        self.clear_dog()
        self.clear_soul()
        self.doppelganger = None
        self.exceptions = set()

    def set_field(self, field):
        self.field = field

    def set_ninja(self, i, y, x):
        self.ninjas[i] = Character(i, y, x)

    def clear_dog(self):
        self.dogs = {}
        self.dog_points = set()

    def set_dog(self, i, y, x):
        self.dogs[Point(y, x)] = Character(i, y, x)
        self.dog_points.add(Point(y, x))

    def clear_soul(self):
        self.souls = set()

    def set_soul(self, y, x):
        self.souls.add(Point(y, x))

    def set_skills(self, skills):
        self.skills = skills

    def dist_to_ninjas(self, point):
        return set([ninja.point.dist(point) for ninja in self.ninjas])

    def get_soul_under_block(self):
        for soul in self.souls:
            if self.field[soul.y][soul.x].is_block:
                return soul

    def dump_field(self):
        for r in range(ROW):
            for c in range(COL):
                w = ' '
                if self.field[r][c].is_wall:
                    w = 'W'
                if self.field[r][c].is_block:
                    w = 'O'
                if Point(r, c) in self.dog_points:
                    w = 'X'
                if Point(r, c) == self.ninjas[0].point:
                    w = '1'
                if Point(r, c) == self.ninjas[1].point:
                    w = '2'
                print(w, end="")
            print("")

    def steps_from_points(self, points):
        class NinjaPoint:
            """
            PriorityQueueに入れる用
            """

            def __init__(self, dist, point):
                self.step = dist
                self.point = point

            def __lt__(self, other):
                if self.step == other.step:
                    return self.point < other.point
                return self.step < other.step

        steps_from_points = [[INF for _ in range(COL)] for _ in range(ROW)]
        visited = set()

        queue = PriorityQueue()
        for point in points:
            queue.put(NinjaPoint(0, point))
            visited.add(point)

        while not queue.empty():
            q = queue.get()
            steps_from_points[q.point.y][q.point.x] = q.step

            for direction in Direction.directions:
                point = q.point + direction
                if not self.field[point.y][point.x].is_empty:
                    continue

                if point not in visited:
                    visited.add(point)
                    queue.put(NinjaPoint(q.step + 1, point))

        return steps_from_points

File: ai/src/state.cpp
#ifndef AI_STATE
#define AI_STATE

#include "state.h"

State::State() : ninjas(NINJA_NUM), nextStep(NINJA_NUM), doppelganger() {
    field = vector<vector<Cell>>(Y);
    REP(i, Y) {
        field[i] = vector<Cell>(X);
    }


}

void State::start() {
    clearDogInfo();
    clearSoul();

    doppelganger = Point();
    exceptions.clear();

    nextStep[0].clear();
    nextStep[1].clear();
}

void State::dumpField(ostream &cerr) {
    Cell cell;
    Point point;
    stringstream ss;
    REP(y, Y) {
        cerr << setw(2) << y;
        REP(x, X) {
            cell = field[y][x];
            point = Point(y, x);
            string c = "";
            if (cell.isWall()) {
                c = "W";
            } else if (cell.isBlock()) {
                c = "O";
            } else if (dogPoints.find(point) != dogPoints.end()) {
                c = "x";
            } else {
                c = "_";
            }

            if (point == ninjas[0].point) {
                c += "1";
            } else if (point == ninjas[1].point) {
                c += "2";
            } else if (point == doppelganger) {
                c += "#";
            } else {
                c += " ";
            }
            cerr << setw(2) << c;
        }
        cerr << endl;
    }
}

set<int> State::distToNinja(const Point &pos) const {
    set<int> dists;
    dists.insert(ninjas[0].point.dist(pos));
    dists.insert(ninjas[1].point.dist(pos));
    return dists;
}

void State::clearDogInfo() {
    dogPoints.clear();
    dogs.clear();
}

void State::clearSoul() {
    souls.clear();
}

void State::setStepsToNinjas() {
    vector<Point> ninjaPoints;
    ninjaPoints.push_back(ninjas[0].point);
    ninjaPoints.push_back(ninjas[1].point);
    stepsToNinjas = stepsFromPoints(ninjaPoints);
}

void State::setStepsToDoppel(Point doppel) {
    vector<Point> points;
    points.push_back(doppel);
    stepsToDopperl = stepsFromPoints(points);
}

void State::setStepsToReachableCellFromNinja() {
    vector<Point> ninjaPoint;

    ninjaPoint.push_back(ninjas[0].point);
    ninjaPoint.push_back(ninjas[1].point);
    stepsToReachableCellNinjas = stepsToReachableCell(ninjaPoint);

}

vector<vector<int>> State::stepsToReachableCell(vector<Point> points) {
    vector<vector<int>> steps(Y);
    // 初期化
    REP (y, Y) {
        steps[y] = vector<int>(X);
        REP(x, X) {
            steps[y][x] = INF;
        }
    }

    set<Point> visited;
    priority_queue<tuple<int, Point>, vector<tuple<int, Point>>, greater<tuple<int, Point>>> queue; //dist, point
    for (Point p : points) {
        queue.push(make_tuple(0, p));
        visited.insert(p);
    }

    while (!queue.empty()) {
        tuple<int, Point> q = queue.top();
        queue.pop();

        int step = get<0>(q);
        Point point = get<1>(q);

        steps[point.y][point.x] = step;

        REP(i, DIRECTION_NUM) {
            Point d = Direction::directions[i];
            Point next = point + d;
            if (field[next.y][next.x].isWall()) continue;
            if (dogPoints.find(next) != dogPoints.end()) continue;          // 犬は壁
            if (field[next.y][next.x].isBlock()) {
                Point nextnext = next + d;
                if (field[nextnext.y][nextnext.x].isBlock()) continue;      // 石石は壁
                if (dogPoints.find(nextnext) != dogPoints.end()) continue;  // 石犬も壁
                //本当は石忍者も壁
                if (field[nextnext.y][nextnext.x].isEmpty()) {              // 石空?は壁
                    Point nnn = nextnext + d;
                    if (!field[nnn.y][nnn.x].isEmpty()) continue;
                    if (dogPoints.find(nnn) != dogPoints.end()) continue;
                }
            }
            if (visited.find(next) == visited.end()) {
                visited.insert(next);
                queue.push(make_tuple(step + 1, next));
            }

        }

    }
    return steps;
}

vector<vector<int>> State::stepsFromPoints(vector<Point> points) {
    vector<vector<int>> steps(Y);
    REP (y, Y) {
        steps[y] = vector<int>(X);
        REP(x, X) {
            steps[y][x] = INF;
        }
    }

    set<Point> visited;
    priority_queue<tuple<int, Point>, vector<tuple<int, Point>>, greater<tuple<int, Point>>> queue; //dist, point
    for (Point p : points) {
        queue.push(make_tuple(0, p));
        visited.insert(p);
    }

    while (!queue.empty()) {
        tuple<int, Point> q = queue.top();
        queue.pop();

        int step = get<0>(q);
        Point point = get<1>(q);

        steps[point.y][point.x] = step;

        REP(i, DIRECTION_NUM) {
            Point d = Direction::directions[i];
            Point next = point + d;
            if (!field[next.y][next.x].isEmpty()) continue;
            if (visited.find(next) == visited.end()) {
                visited.insert(next);
                queue.push(make_tuple(step + 1, next));
            }

        }

    }
    return steps;
}

void State::dumpStepsToNinjas(ostream &cerr) {

    REP(y, Y) {
        REP(x, X) {
            cerr << setw(2);

            if (stepsToNinjas[y][x] == INF) {
                cerr << "__";
            } else if (souls.find(Point(y, x)) != souls.end()) {
                cerr << "@";
            } else {
                cerr << stepsToNinjas[y][x];
            }
            cerr << " ";
        }
        cerr << endl;
    }
}

void State::dumpStepsToDoppel(ostream &cerr) {

    REP(y, Y) {
        REP(x, X) {
            cerr << setw(2);

            if (stepsToDopperl[y][x] == INF) {
                cerr << "__";
            } else if (souls.find(Point(y, x)) != souls.end()) {
                cerr << "@";
            } else {
                cerr << stepsToDopperl[y][x];
            }
            cerr << " ";
        }
        cerr << endl;
    }
}


#endif